test/libsolidity/GasTest.cpp

```diff
@@ -0,0 +1,222 @@
+			skipWhitespace(it, line.end());
+			if (*it++ != '+')
+				BOOST_THROW_EXCEPTION(runtime_error("Invalid expectation: expected \"+\"-"));
+			skipWhitespace(it, line.end());
+			m_creationCost.codeDepositCost = parseGasCost(it, line.end());
+			skipWhitespace(it, line.end());
+			if (*it++ != '=')
+				BOOST_THROW_EXCEPTION(runtime_error("Invalid expectation: expected \"+\"-"));
+			skipWhitespace(it, line.end());
+			m_creationCost.totalCost = parseGasCost(it, line.end());
+		}
+		else if (line == "// external:")
+			currentKind = &m_externalFunctionCosts;
+		else if (line == "// internal:")
+			currentKind = &m_internalFunctionCosts;
+		else if (!currentKind)
+			BOOST_THROW_EXCEPTION(runtime_error("No function kind specified. Expected \"external:\" or \"internal:\"."));
+		else
+		{
+			if (!boost::starts_with(line, "// "))
+				BOOST_THROW_EXCEPTION(runtime_error("Invalid expectation: expected \"// \"."));
+			auto it = line.begin() + 3;
+			skipWhitespace(it, line.end());
+			auto functionNameBegin = it;
+			while (it != line.end() && *it != ':')
+				++it;
+			std::string functionName(functionNameBegin, it);
+			if (functionName == "fallback")
+				functionName.clear();
+			expect(it, line.end(), ':');
+			skipWhitespace(it, line.end());
+			if (it == line.end())
+				BOOST_THROW_EXCEPTION(runtime_error("Invalid expectation: expected gas cost."));
+			(*currentKind)[functionName] = std::string(it, line.end());
+		}
+}
+
+void GasTest::printUpdatedExpectations(std::ostream& _stream, std::string const& _linePrefix) const
+{
+	Json::Value estimates = compiler().gasEstimates(compiler().lastContractName());
+	_stream << _linePrefix << "creation: "
+			<< estimates["creation"]["executionCost"].asString() << " + "
+			<< estimates["creation"]["codeDepositCost"].asString() << " = "
+			<< estimates["creation"]["totalCost"].asString() << std::endl;
+
+	for (auto kind: {"external", "internal"})
+		if (estimates[kind])
+		{
+			_stream << _linePrefix << kind << ":" << std::endl;
+			for (auto it = estimates[kind].begin(); it != estimates[kind].end(); ++it)
+			{
+				_stream << _linePrefix << "  ";
+				if (it.key().asString().empty())
+					_stream << "fallback";
+				else
+					_stream << it.key().asString();
+				_stream << ": " << it->asString() << std::endl;
+			}
+		}
+}
+
+
+bool GasTest::run(ostream& _stream, string const& _linePrefix, bool)
+{
+	string const versionPragma = "pragma solidity >=0.0;\n";
+	compiler().reset();
+	OptimiserSettings settings = m_optimise ? OptimiserSettings::standard() : OptimiserSettings::minimal();
+	if (m_optimiseYul)
+	{
+		settings.runYulOptimiser = m_optimise;
+		settings.optimizeStackAllocation = m_optimise;
+	}
+	settings.expectedExecutionsPerDeployment = m_optimiseRuns;
+	compiler().setOptimiserSettings(settings);
+	compiler().setSources({{"", versionPragma + m_source}});
+
+	if (compiler().parse())
+		if (compiler().analyze())
+			compiler().compile();
```

Comment:
What if it fails?

Reply:
``gasEstimates`` throws in that case.

Reply:
It might be nice to show the compiler error messages, though...

Reply:
Ok!